core/DataAccess/LogAggregator.php

```diff
@@ -174,18 +187,107 @@ public function setQueryOriginHint($nameOfOrigiin)
         $this->queryOriginHint = $nameOfOrigiin;
     }
 
+    private function getSegmentTmpTableName()
+    {
+        $bind = $this->getGeneralQueryBindParams();
+        return self::LOG_TABLE_SEGMENT_TEMPORARY_PREFIX . md5(json_encode($bind) . $this->segment->getString());
+    }
+
+    public function cleanup()
+    {
+        if (!$this->segment->isEmpty() && $this->isSegmentCacheEnabled()) {
+            $segmentTable = $this->getSegmentTmpTableName();
+            $segmentTable = Common::prefixTable($segmentTable);
+            Db::getReader()->query('DROP TABLE IF EXISTS ' . $segmentTable);
+
+            $logTablesProvider = $this->getLogTableProvider();
+            if ($logTablesProvider->getLogTable($segmentTable)) {
+                $logTablesProvider->setTempTable(null); // no longer available
+            }
+        }
+    }
+
+    private function isSegmentCacheEnabled()
+    {
+        $config = Config::getInstance();
+        $general = $config->General;
+        return !empty($general['enable_segments_cache']);
+    }
+
+    private function getLogTableProvider()
+    {
+        return StaticContainer::get(LogTablesProvider::class);
+    }
+
     public function generateQuery($select, $from, $where, $groupBy, $orderBy, $limit = 0, $offset = 0)
     {
+        $segment = $this->segment;
         $bind = $this->getGeneralQueryBindParams();
-        $query = $this->segment->getSelectQuery($select, $from, $where, $bind, $orderBy, $groupBy, $limit, $offset);
+
+        if (!$this->segment->isEmpty() && $this->isSegmentCacheEnabled()) {
+            // todo when we create table for query we cannot know if the archiver will use that same Db writer or Db reader
+            // if plugin doesn't support reader just yet... there will be a problem!
```

Comment:
Does this TODO need to be turned into an issue for followup?

Reply:
Not needed anymore. We discussed to document this.